### PR TITLE
Add chunks to dask.array.broadcast_to

### PR DESCRIPTION
**Pull request by shoyer**
This allows for directly producing output arrays with proper chunking for broadcasting, instead of relying on rechunking afterwards. As mentioned by @magonser in https://github.com/dask/dask/issues/2057#issuecomment-333332368 the later can be slower, because broadcast arrays get fully loaded into memory by indexing.

On one toy benchmark, I see an improvement from 6.5 to 4.0 seconds:
```
In [12]: x = da.ones((1, 1000,), chunks=10)

In [13]: y = da.ones((1000, 1), chunks=10)

In [14]: z = x + y
/Users/shoyer/dev/dask/dask/array/core.py:2113: UserWarning: Increasing number of chunks by factor of 100
  (nparts / max_parts))

In [15]: %timeit z.compute()
1 loop, best of 3: 1.3 s per loop

In [16]: z = da.broadcast_to(x, (1000, 1000)) + da.broadcast_to(y, (1000, 1000))
/Users/shoyer/dev/dask/dask/array/core.py:2113: UserWarning: Increasing number of chunks by factor of 100
  (nparts / max_parts))

In [17]: %timeit z.compute()
1 loop, best of 3: 6.55 s per loop

In [18]: z = da.broadcast_to(x, (1000, 1000), 10) + da.broadcast_to(y, (1000, 1000), 10)

In [19]: %timeit z.compute()
1 loop, best of 3: 3.97 s per loop
```

This isn't as much as a I hoped, but it's something. Maybe there are other cases where it makes more of a difference.

- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -1850,6 +1851,19 @@ def normalize_chunks(chunks, shape=None):
...
+def normalize_and_check_chunks(chunks, shape):
+    """Normalize chunks and check them for consistency."""
+    chunks = normalize_chunks(chunks, shape)
+    if len(chunks) != len(shape):
+        raise ValueError("Input array has %d dimensions but the supplied "
+                         "chunks has only %d dimensions" %
+                         (len(shape), len(chunks)))
+    if tuple(map(sum, chunks)) != shape:
+        raise ValueError("Chunks do not add up to shape. "
+                         "Got chunks=%s, shape=%s" % (chunks, shape))
```
Comment (mrocklin) [inline dask/array/core.py]: Should this check just be added to `normalize_chunks` instead?

File: dask/array/core.py [hunk 2]
```diff
@@ -2650,12 +2664,34 @@ def _enforce_dtype(*args, **kwargs):
...
+    See Also
+    --------
+    numpy.broadcast_to
+    """
     x = asarray(x)
     shape = tuple(shape)
 
-    if x.shape == shape:
+    if x.shape == shape and (chunks is None or chunks == x.chunks):
         return x
```
Comment (jakirkham) [inline dask/array/core.py]: So I guess this means that if the chunks are different and the shape is the same we proceed. What about warning and doing nothing, or erroring instead?

Comment (jakirkham) [reply dask/array/core.py]: Adding to that thought, we might want to consider what it means if `x.chunks` mismatches with `chunks[:x.ndim]`. Should this generally be giving a warning, error, or quietly rechunking the data under the hood? Likely with an exemption for dimensions of length `1`.

Comment (jakirkham) [reply dask/array/core.py]: NVM I guess this is handled below. Maybe it would make sense to move that check up here so it is more visible. Not sure if that is straightforward.

File: dask/array/core.py (hunk 1 above)
Comment (shoyer) [reply dask/array/core.py]: The checks currently aren't done in the `Array` constructor, which I assumed was an optimization to make creating an `Array` with internal dask.array functions less expensive. But if this doesn't meaningfully slow things down maybe it should be revisited.

File: dask/array/core.py (hunk 2 above)
Comment (shoyer) [reply dask/array/core.py]: I think this exists to avoid doing all the expensive validation if it isn't necessary.

File: dask/array/core.py (hunk 1 above)
Comment (mrocklin) [reply dask/array/core.py]: Timing for reference

```python
In [1]: chunks = ((1000,) * 100, (1000,) * 200)

In [2]: shape = (100000, 100000)

In [3]: %timeit tuple(map(sum, chunks)) != shape
The slowest run took 7.67 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 2.78 µs per loop
```

This seems fine to me.

**Comment by magonser** [comment]
tldr; thanks @shoyer for this PR! It is appreciated (by me) a lot. You mentioned speed gain is not as much as you hoped for; my experience has been that once switched to a configuration with distributed scheduler and multiple workers, performance gains are better with this new `broadcast_to`. 

PS: Also thanks @jakirkham for PR #2938, after this PR is done, `meshgrid` maybe could be adjusted to the use this way of chunking.

**Comment by jakirkham** [comment]
> Also thanks @jakirkham for PR #2938, after this PR is done, `meshgrid` maybe could be adjusted to the use this way of chunking. 

np. Not sure I follow the suggestion @magonser. `meshgrid` was designed to just use the chunking of the arrays provided as input.

**Comment by jakirkham** [comment]
Other than the concerns already raised, this seems fine to me.

**Comment by shoyer** [comment]
@jakirkham The advantage of using `broadcast_to` for meshgrid is that broadcasting is faster than repeating -- it doesn't require copying any data, just adjusting strides/shape.

**Comment by jakirkham** [comment]
If it improves the performance, I have no objections. Would happily review such a PR.

**Comment by magonser** [comment]
I've done some benchmarking also for larger than memory problems (with final reduction stages before the end result) and this PR improves performance for my use cases. 

@jakirkham: yes, `meshgrid` should not accept any `chunks` arguments, however it should use this PRs `broadcast_to` or at least it's strategy underneath. At the moment `meshgrid` produces rechunking operations in the DAG.

**Comment by shoyer** [comment]
Test failures seem to be unrelated: `test_local_get_with_distributed_active`

**Comment by jakirkham** [comment]
Yeah have been seeing similar failures in other CI builds as well. 😕

**Comment by jakirkham** [comment]
Should be fixed after PR ( https://github.com/dask/dask/pull/2989 ) lands. Will restart after.

**Comment by jakirkham** [comment]
Toggling for CI.

**Comment by jakirkham** [comment]
Thanks @shoyer.

Side note: Sorry for the slow close/reopen. Thought I had reopened it earlier and only now realized I hadn't.
